src/java.base/share/classes/sun/util/locale/provider/CalendarNameProviderImpl.java

```diff
@@ -189,7 +189,20 @@ public String getDisplayNameImpl(String calendarType, int field, int value, int
                             if (name.isEmpty()) {
                                 continue;
                             }
-                            map.put(name, base + i);
+                            if (field == AM_PM) {
+                                // Convert the dayPeriod "type" to either AM or PM field value.
+                                // Following is a mapping of dayPeriod "type" to their array index
+                                // in Calendar resource bundle
+                                // am = 0, pm = 1, midnight = 2, noon = 3, morning1 = 4,
+                                // morning2 = 5, afternoon1 = 6, afternoon2 = 7, evening1 = 8,
+                                // evening2 = 9, night1 = 10, night2 = 11
+                                switch (i) {
+                                    case 0, 2, 3, 4, 5 -> map.put(name, AM);
+                                    case 1, 6, 7, 8, 9, 10, 11 -> map.put(name, PM);
```

Comment:
One part where I need input here, is the "midnight" and the "noon" types. I haven't found anything in the JDK code which I can use as a reference to classify "midnight" and "noon" as `AM` (which is what this code is doing). The doc here https://unicode.org/reports/tr35/tr35-dates.html#Day_Period_Rules speaks about these 2 types but I don't think it's something that I can use to translate those types into a field value for Calendar, to represent `AM` or `PM`. As such, right now, I've just guessed that `AM` is probably the right value for these types, but there's no technical reason or reference backing it.